zcash_client_sqlite/src/wallet.rs

```diff
@@ -808,213 +808,442 @@ pub(crate) fn get_derived_account<P: consensus::Parameters>(
+                let last_completed_subtree = conn
+                    .query_row(
+                        &format!(
+                            "SELECT shard_index, subtree_end_height
+                            FROM {table_prefix}_tree_shards
+                            WHERE subtree_end_height IS NOT NULL
+                            ORDER BY shard_index DESC
+                            LIMIT 1"
+                        ),
+                        [],
+                        |row| Ok((row.get::<_, u64>(0)?, BlockHeight::from_u32(row.get(1)?))),
+                    )
+                    // `None` if we have no subtree roots yet.
+                    .optional()?;
 
-            // We don't have complete information on how many outputs will exist in the shard at
-            // the chain tip without having scanned the chain tip block, so we overestimate by
-            // computing the maximum possible number of notes directly from the shard indices.
-            //
-            // TODO: it would be nice to be able to reliably have the size of the commitment tree
-            // at the chain tip without having to have scanned that block.
-            Ok(conn
-                .query_row(
-                    "SELECT MIN(shard_index), MAX(shard_index)
-                     FROM sapling_tree_shards
-                     WHERE subtree_end_height > :start_height
-                     OR subtree_end_height IS NULL",
-                    named_params![":start_height": u32::from(birthday_height)],
-                    |row| {
-                        let min_tree_size = row
-                            .get::<_, Option<u64>>(0)?
-                            .map(|min_idx| min_idx << SAPLING_SHARD_HEIGHT);
-                        let max_tree_size = row
-                            .get::<_, Option<u64>>(1)?
-                            .map(|max_idx| (max_idx + 1) << SAPLING_SHARD_HEIGHT);
-                        Ok(start_size.or(min_tree_size).zip(max_tree_size).map(
-                            |(min_tree_size, max_tree_size)| {
-                                Ratio::new(
-                                    scanned_count.unwrap_or(0),
-                                    max_tree_size - min_tree_size,
+                if let Some((last_completed_subtree_index, last_completed_subtree_height)) =
+                    last_completed_subtree
+                {
+                    // If we know the tree size at the last scanned height, and that
+                    // height is within the incomplete subtree, extrapolate.
+                    let tip_tree_size =
+                        last_scanned.and_then(|(last_scanned, last_scanned_tree_size)| {
+                            (last_scanned > last_completed_subtree_height)
+                                .then(|| {
+                                    let scanned_notes = last_scanned_tree_size
+                                        - (last_completed_subtree_index << shard_height);
```

Comment:
Shouldn't this be `(last_completed_subtree_index + 1) << shard_height`? Without the `+1` this computes the start of a subtree, not the end as the indices are zero-based.

For a moment I thought you were averaging over all the data to the start of the previous subtree, but that's not the case as you're using the subtree end height.

Reply:
It would be clearer in this code to name `last_completed_subtree_height` to `last_completed_subtree_end` or `incomplete_subtree_start_height`.

Reply:
Another possibility is to use the `Address` for the root of the subtree. The whole reason that type exists is to make errors like this more obvious.